File: clients/java/exp-client/src/main/java/CAC/ExperimentationClient.java
package CAC;

import java.io.IOException;

import jnr.ffi.LibraryLoader;
import jnr.ffi.Pointer;


public class ExperimentationClient {
    public interface RustLib {
        int expt_new_client(String tenant, long updateFrequency, String hostName);

        void expt_free_client(Pointer ptr);

        Pointer expt_get_client(String tenant);

        String expt_last_error_message();

        void expt_free_string(Pointer s);

        void expt_start_polling_update(String tenant);

        Pointer expt_get_applicable_variant(Pointer clientPtr, String context, short toss);

        Pointer expt_get_satisfied_experiments(Pointer clientPtr, String context, String filter_prefix);

        Pointer expt_get_filtered_satisfied_experiments(Pointer clientPtr, String context, String filter_prefix);

        Pointer expt_get_running_experiments(Pointer clientPtr);
    }

    public static RustLib rustLib;

    public ExperimentationClient(String libraryPath, String libraryName) {
        System.setProperty("jnr.ffi.library.path", libraryPath);
        ExperimentationClient.rustLib = LibraryLoader.create(RustLib.class).load(libraryName);
    }

    public int exptNewClient(String tenant, long updateFrequency, String hostName) throws IOException {
        int result = rustLib.expt_new_client(tenant, updateFrequency, hostName);
        if (result > 0) {
            String errorMessage = rustLib.expt_last_error_message();
            throw new IOException("Failed to create new Experimentation client: " + errorMessage);
        }
        return result;
    }

    public Pointer getExptClient(String tenant) throws IOException {
        Pointer clientPtr = rustLib.expt_get_client(tenant);
        return clientPtr;
    }

    public void startPollingUpdate(String tenant) throws IOException {
        rustLib.expt_start_polling_update(tenant);
    }

    public String getApplicableVariants (Pointer clientPtr, String context, short toss) throws IOException {
        Pointer result = rustLib.expt_get_applicable_variant(clientPtr, context, toss);
        if (result == null) {
            String errorMessage = rustLib.expt_last_error_message();
            throw new IOException("Failed to get applicable variants for the experiment: " + errorMessage);
        }
        String applicableVariants = getStringAndFree(result);
        return applicableVariants;
    }

    public String getSatisfiedVariants(Pointer clientPtr, String context, String filter_prefix) throws IOException {
        Pointer result = rustLib.expt_get_satisfied_experiments(clientPtr, context, filter_prefix);
        if (result == null) {
            String errorMessage = rustLib.expt_last_error_message();
            throw new IOException("Failed to get satisfied variants for the experiment: " + errorMessage);
        }
        String satisfiedVariants = getStringAndFree(result);
        return satisfiedVariants;
    }

    public String getFilteredSatisfiedVariants(Pointer clientPtr, String context, String filter_prefix) throws IOException {
        Pointer result = rustLib.expt_get_filtered_satisfied_experiments(clientPtr, context, filter_prefix);
        if (result == null) {
            String errorMessage = rustLib.expt_last_error_message();
            throw new IOException("Failed to get filtered satisfied variants for the experiment: " + errorMessage);
        }
        String filteredSatisfiedVariants = getStringAndFree(result);
        return filteredSatisfiedVariants;
    }

    public String getRunningExperiments(Pointer clientPtr) throws IOException {
        Pointer result = rustLib.expt_get_running_experiments(clientPtr);
        if (result == null) {
            String errorMessage = rustLib.expt_last_error_message();
            throw new IOException("Failed to get the running variants for the experiment: " + errorMessage);
        }
        String runningExperiments = getStringAndFree(result);
        return runningExperiments;
    }

    public String getLastError() throws IOException {
        String errorMessage = rustLib.expt_last_error_message();
        if (errorMessage != null) {
            return errorMessage;
        } else {
            return "No error";
        }
    }

    public void freeString(Pointer ptr) {
        rustLib.expt_free_string(ptr);
    }

    public String getStringAndFree(Pointer ptr) {
        if (ptr == null) {
            return null;
        }
        try {
            return ptr.getString(0);
        } finally {
            freeString(ptr);
        }
    }
}

File: clients/lua/expclient/ExperimentationClient.lua
local ffi = require("ffi")
local platform = string.lower(jit.os)

local lib_path = os.getenv("SUPERPOSITION_LIB_PATH")
if not lib_path then
    error("Environment variable SUPERPOSITION_LIB_PATH is not set")
end

if platform == "osx" then
    lib_path = lib_path .. "/libexperimentation_client.dylib"
elseif platform == "linux" then
    lib_path = lib_path .. "/libexperimentation_client.so"
elseif platform == "windows" then
    lib_path = lib_path .. "/libexperimentation_client.dll"
else
    error("Unsupported platform: " .. platform)
end

ffi.cdef[[
    int expt_new_client(const char* tenant_name, int polling_frequency, const char* cac_host_name);
    void expt_start_polling_update(const char* tenant_name);
    const char* expt_get_client(const char* tenant_name);
    const char* expt_get_applicable_variant(const char* client_ptr, const char* context, int toss);
    const char* expt_get_satisfied_experiments(const char* client_ptr, const char* context, const char* filter_prefix);
    const char* expt_get_filtered_satisfied_experiments(const char* client_ptr, const char* context, const char* filter_prefix);
    const char* expt_get_running_experiments(const char* client_ptr);
    void expt_free_string(const char* string);
    const char* expt_last_error_message();
    int expt_last_error_length();
    void expt_free_client(const char* client_ptr);
]]

local rust_lib = ffi.load(lib_path)

local ExperimentationClient = {}
ExperimentationClient.__index = ExperimentationClient

function ExperimentationClient:new(tenant_name, polling_frequency, cac_host_name)
    assert(tenant_name and #tenant_name > 0, "tenantName cannot be null or empty")
    assert(cac_host_name and #cac_host_name > 0, "cacHostName cannot be null or empty")

    local self = setmetatable({}, ExperimentationClient)
    self.tenant = tenant_name
    self.polling_frequency = polling_frequency
    self.cac_host_name = cac_host_name
    return self
end

function ExperimentationClient:get_experimentation_last_error_message()
    local error_message = rust_lib.expt_last_error_message();
    if error_message == nil then
        return "No Error"
    end
    return ffi.string(rust_lib.expt_last_error_message())
end

function ExperimentationClient:create_new_experimentation_client()
    local resp_code = rust_lib.expt_new_client(self.tenant, self.polling_frequency, self.cac_host_name)
    if resp_code == 1 then
        local error_message = self:get_experimentation_last_error_message()
        print("Some error occurred while creating new experimentation client:", error_message)
        error("Client Creation Error")
    end
    return resp_code
end

function ExperimentationClient:get_experimentation_client()
    return ffi.string(rust_lib.expt_get_client(self.tenant))
end

function ExperimentationClient:get_running_experiments()
    return ffi.string(rust_lib.expt_get_running_experiments(self:get_experimentation_client()))
end

function ExperimentationClient:free_string(string)
    rust_lib.expt_free_string(string)
end

function ExperimentationClient:start_experimentation_polling_update()
    rust_lib.expt_start_polling_update(self.tenant)
end

function ExperimentationClient:get_experimentation_last_error_length()
    return rust_lib.expt_last_error_length()
end

function ExperimentationClient:free_experimentation_client()
    rust_lib.expt_free_client(self:get_experimentation_client())
end

function ExperimentationClient:get_filtered_satisfied_experiments(context, filter_prefix)
    local resp = rust_lib.expt_get_filtered_satisfied_experiments(self:get_experimentation_client(), context, filter_prefix)
    if resp == nil then
        return ""
    end
    return ffi.string(resp)
end

function ExperimentationClient:get_applicable_variant(context, toss)
    return ffi.string(rust_lib.expt_get_applicable_variant(self:get_experimentation_client(), context, toss))
end

function ExperimentationClient:get_satisfied_experiments(context, filter_prefix)
    return ffi.string(rust_lib.expt_get_satisfied_experiments(self:get_experimentation_client(), context, filter_prefix))
end

return ExperimentationClient